lib/internal/repl/history.js

```diff
@@ -17,168 +25,405 @@ let debug = require('internal/util/debuglog').debuglog('repl', (fn) => {
 });
 const permission = require('internal/process/permission');
 const { clearTimeout, setTimeout } = require('timers');
+const {
+  reverseString,
+} = require('internal/readline/utils');
 
-const noop = FunctionPrototype;
-
-// XXX(chrisdickinson): The 15ms debounce value is somewhat arbitrary.
 // The debounce is to guard against code pasted into the REPL.
 const kDebounceHistoryMS = 15;
+const kHistorySize = 30;
 
-module.exports = setupHistory;
+// Class fields
+const kTimer = Symbol('_kTimer');
+const kWriting = Symbol('_kWriting');
+const kPending = Symbol('_kPending');
+const kRemoveHistoryDuplicates = Symbol('_kRemoveHistoryDuplicates');
+const kHistoryHandle = Symbol('_kHistoryHandle');
+const kHistoryPath = Symbol('_kHistoryPath');
+const kContext = Symbol('_kContext');
+const kIsFlushing = Symbol('_kIsFlushing');
+const kHistory = Symbol('_kHistory');
+const kSize = Symbol('_kSize');
+const kIndex = Symbol('_kIndex');
 
-function _writeToOutput(repl, message) {
-  repl._writeToOutput(message);
-  repl._refreshLine();
-}
+// Class methods
+const kNormalizeLineEndings = Symbol('_kNormalizeLineEndings');
+const kWriteToOutput = Symbol('_kWriteToOutput');
+const kOnLine = Symbol('_kOnLine');
+const kOnExit = Symbol('_kOnExit');
+const kInitializeHistory = Symbol('_kInitializeHistory');
+const kHandleHistoryInitError = Symbol('_kHandleHistoryInitError');
+const kHasWritePermission = Symbol('_kHasWritePermission');
+const kValidateOptions = Symbol('_kValidateOptions');
+const kResolveHistoryPath = Symbol('_kResolveHistoryPath');
+const kReplHistoryMessage = Symbol('_kReplHistoryMessage');
+const kFlushHistory = Symbol('_kFlushHistory');
+const kGetHistoryPath = Symbol('_kGetHistoryPath');
 
-function setupHistory(repl, historyPath, ready) {
-  // Empty string disables persistent history
-  if (typeof historyPath === 'string')
-    historyPath = StringPrototypeTrim(historyPath);
+class ReplHistory {
+  constructor(context, options) {
+    this[kValidateOptions](options);
 
-  if (historyPath === '') {
-    repl._historyPrev = _replHistoryMessage;
-    return ready(null, repl);
+    this[kHistoryPath] = ReplHistory[kGetHistoryPath](options);
+    this[kContext] = context;
+    this[kTimer] = null;
+    this[kWriting] = false;
+    this[kPending] = false;
+    this[kRemoveHistoryDuplicates] = options.removeHistoryDuplicates || false;
+    this[kHistoryHandle] = null;
+    this[kIsFlushing] = false;
+    this[kSize] = options.historySize ?? context.historySize ?? kHistorySize;
+    this[kHistory] = options.history ?? [];
+    this[kIndex] = -1;
   }
 
-  if (!historyPath) {
-    try {
-      historyPath = path.join(os.homedir(), '.node_repl_history');
-    } catch (err) {
-      _writeToOutput(repl, '\nError: Could not get the home directory.\n' +
-        'REPL session history will not be persisted.\n');
+  initialize(onReadyCallback) {
+    // Empty string disables persistent history
+    if (this[kHistoryPath] === '') {
```

Comment:
if the user provided an object without a `historyFile` then I believe that `this[kHistoryPath]` would be `undefined` here, am I correct? what would the meaning of that be? should history also be disabled in that case?

related, I haven't notice it anywhere, should `historyFile` be required (and some validation put in place to make sure that the property is on the history config object?)

Reply:
In case the `historyFile` is undefined, it is fallbacking to the standard path `~/.node_repl_history`, so it should not be required 😄